### DISCUSSION TITLE
OPA-Istio Kubernetes - Does the sidecar OPA read from the admission controller cache?

### DISCUSSION THREAD

**Discussion by Thomas-Rice** [open-policy-agent/community#272] [Envoy]
Hi,
I've injected data into the admission controller, but when i create a new deployment with a sidecar OPA, I cannot seem to get the data that is stored in the admission controllers cache. 
Would anyone know how this works? If the sidecars are meant to be their own isolated OPA or talk to the central admission controller for the data?

**Comment by anderseknert** [comment]
Hi Thomas! 🙂 I'm not sure what you are referring to as an admission controller here. Are you using OPA for Kubernetes admission control as well? That deployment would normally be kept separate from any other type of OPA deployments in you cluster. What data from the admission controller would you want to make available in you sidecars?

**Reply by Thomas-Rice** [reply]
Hi thanks for the reply! 
Sorry I forgot to post the setup that I followed! 

I used the following setup from the quick_start.yaml to get OPA/Istio/Envoy in my cluster
https://github.com/open-policy-agent/opa-envoy-plugin/blob/main/examples/istio/quick_start.yaml

**_Are you using OPA for Kubernetes admission control as well?_**
That's correct, sorry I didn't explain that! 

So we have the admission controller in the opa-istio namespace and have pushed data to that via it's api. 

In my case, when create a new namespace for a api that I have made and enable the OPA sidecar injection. 
When I'm trying to read the data in a policy rego file, it looks like the policy can't read that data, however i'm not sure on this and not sure how to test as accessing the sidecar is a bit awkward.

Would you know if the sidecar is meant to be able to read the data from the opa-istio namespace or if there is additional setup I need to do if it's possible?

**Reply by anderseknert** [reply]
Each OPA has it's own in-memory store, and there's no synchronization of that between OPAs in a cluster. The preferred (IMHO, of course) way to provide your sidecars with data would be via the bundle API, but if you have some other way to provide data to your admission controller OPA, perhaps that could be extended to provide the same data for the OPA sidecars that have a need for it.

**Reply by Thomas-Rice** [reply]
Ok fantastic,
I really appreciate you clearing that up for me :) 

So just to clarify, 
I'm already reading my policy from an S3 bucket, so if I bundle my data with that, then the sidecars will be able to read that data?

**Reply by anderseknert** [reply]
Indeed! You could just configure your sidecar OPAs to either fetch the same bundle as your admission controller, or you could build a bundle for your sidecars with only policy and data relevant for those. I'd probably go with the latter in order to isolate the two from each other, even if the data will be similar / identical.

**Reply by Thomas-Rice** [reply]
Thank you so much for the advice!

**Reply by Thomas-Rice** [reply]
Are there any examples on how to get a sidecar to read a specific bundle?

**Reply by anderseknert** [reply]
You'd just start it with a configuration pointing to any given bundle at a remote endpoint: https://www.openpolicyagent.org/docs/latest/configuration/#example
